docs/Data-flow-in-Site-Kit.md

```diff
@@ -1,3 +1,155 @@
+2. REST API routes (`get_datapoint_definitions`)
+3. Data request to make to a 3rd-party API, if needed (`create_data_request`)
+4. Response parsing for the optional data request (`parse_data_response`)
+
+### Set up remote requests
+
+Each module can set up the services that will be used to make third-party requests. The `setup_services` method returns an array of service identifier to service object instances. These "service objects" can be used to interact with a third-party API.
+
+The identifier is referenced in the datapoint definitions in `get_datapoint_definitions` and the service instance when making the request in `create_data_request`. Here is an example `setup_services` implementation from the Search Console module.
+
+```php
+  protected function setup_services( Google_Site_Kit_Client $client ) {
+    return array(
+      'searchconsole' => new Google_Service_SearchConsole( $client ),
+    );
+  }
+```
+
+### REST API routes
+
+Define a module's REST routes using the `get_datapoint_definitions` function. This function defines the array of REST routes using an array where the keys are the REST routes and the values are requirements/properties of that route. Here's an example from Search Console's `get_datapoint_definitions`:
+
+```php
+protected function get_datapoint_definitions() {
+  return array(
+   'GET:matched-sites'   => array( 'service' => 'searchconsole' ),
+   'GET:searchanalytics' => array(
+    // The `service` key means this request requires Search Console to be
+    // active/connected.
+    'service'   => 'searchconsole',
+    // Shareable means this REST route is shared with Dashboard Sharing users.
+    'shareable' => true,
+   ),
+   'POST:site'           => array( 'service' => 'searchconsole' ),
+   'GET:sites'           => array( 'service' => 'searchconsole' ),
+  );
+ }
+```
+
+### Remote API requests
+
+If a REST route should communicate with a third-party API, eg. a Google API, that should be defined in the `create_data_request` function for a `Module`. This is usually written as a `switch` case, with each route parsing request params, user info, etc. before making the HTTP request to the remote API from PHP.
+
+This code will be used to create the request sent to the third-party API in a standardized way.
+
+Here's an example from the Tag Manager module's `create_data_request` function:
+
+```php
+protected function create_data_request( Data_Request $data ) {
+  switch ( "{$data->method}:{$data->datapoint}" ) {
+    case 'GET:accounts':
+    case 'GET:accounts-containers':
+      return $this->get_tagmanager_service()->accounts->listAccounts();
+  }
+
+  return parent::create_data_request( $data );
+}
+```
+
+### Response parsing for data requests
+
+Any custom data handling, code to run after receiving data, etc. can be defined in the `parse_data_response` method. At its most basic, it can return the response from the remote API without modification or any other action:
```

Comment:
It's worth noting that our convention is to return the subject of the remote resource as the top-level response from our API here rather than passing the response through verbatim. E.g. `listAccounts` would return the array of accounts to the consumer but the actual response from the Google API would include other top-level data (e.g. pagination) that we don't use.